## Entity Matching Notebook for Stage 3 of Data Science Project
### Game Website Dataset from Gamespot and howLongToBeat

In [33]:
import py_entitymatching as em
import pandas as pd
import sys
import math
from datetime import datetime

# Display the versions
print('python version: ' + sys.version )
print('pandas version: ' + pd.__version__ )
print('magellan version: ' + em.__version__ )

python version: 3.6.1 |Anaconda custom (64-bit)| (default, May 11 2017, 13:25:24) [MSC v.1900 64 bit (AMD64)]
pandas version: 0.20.1
magellan version: 0.2.0


#### Reading Input dataset

In [3]:
#A - Gamespot.csv
#B - howLongToBeat.csv

dfTableA = em.read_csv_metadata('../data/Gamespot.csv', key='ID')
dfTableB = em.read_csv_metadata('../data/howLongToBeat.csv', key='ID')

Metadata file is not present in the given path; proceeding to read the csv file.
Metadata file is not present in the given path; proceeding to read the csv file.


In [4]:
print('Number of tuples in A: ' + str(len(dfTableA)))
print('Number of tuples in B: ' + str(len(dfTableB)))
print('Number of tuples in A X B (i.e the cartesian product): ' + str(len(dfTableA)*len(dfTableB)))

Number of tuples in A: 8884
Number of tuples in B: 8481
Number of tuples in A X B (i.e the cartesian product): 75345204


In [5]:
dfTableA.head()

ID                          Title      Developer  \
0  a1    .hack//G.U. vol. 1//Rebirth  CyberConnect2   
1  a2  .hack//G.U. vol. 2//Reminisce  CyberConnect2   
2  a3        .hack//Infection Part 1  CyberConnect2   
3  a4         .hack//Mutation Part 2  CyberConnect2   
4  a5         .hack//Outbreak Part 3  CyberConnect2   

                   Publisher       Platform                Genre  ReleaseDate  \
0  Namco Bandai Games|Bandai  PlayStation 2  Action|Role-Playing  24 Oct 2006   
1  Bandai|Namco Bandai Games  PlayStation 2  Action|Role-Playing   8 May 2007   
2                     Bandai  PlayStation 2  Action|Role-Playing  10 Feb 2003   
3                     Bandai  PlayStation 2  Role-Playing|Action   6 May 2003   
4                     Bandai  PlayStation 2  Action|Role-Playing   9 Sep 2003   

   Rating  CleanRDate  
0     8.8  2006-10-24  
1     8.9  2007-05-08  
2     7.9  2003-02-10  
3     7.9  2003-05-06  
4     8.0  2003-09-09

In [6]:
dfTableB.head()

ID                         Title          Developer          Publisher  \
0  b1                   007 Legends            Eurocom         Activision   
1  b2  007: The World Is Not Enough  Eurocom|Black Ops    Electronic Arts   
2  b3                     0RBITALIS       Alan Zucconi                NaN   
3  b4                      10000000  EightyEight Games  EightyEight Games   
4  b5                     1000 Amps     Brandon Brizzi     Brandon Brizzi   

                          Platform                      Genre  \
0  PC|PlayStation 3|Wii U|Xbox 360       First-Person|Shooter   
1          Nintendo 64|PlayStation       First-Person|Shooter   
2                               PC                     Puzzle   
3             Android|iOS|Linux|PC               Role-Playing   
4                               PC  Adventure|Platform|Puzzle   

        ReleaseDate  Rating  CleanRDate  
0   16 October 2012     5.0  2012-10-16  
1   17 October 2000     7.0  2000-10-17  
2     16 April 2014     8.0  2014-04-16  
3    29 August 2012     7.7  2012-08-29  
4  22 February 2012     6.9  2012-02-22

### Blocking Table to get Candidate Set

#### Try Overlap Blocker
Starting with q-gram overlap

In [7]:
ob = em.OverlapBlocker()

In [17]:
C1 = ob.block_tables(dfTableA, dfTableB, 'Title', 'Title', q_val=3, word_level=False, overlap_size=4, rem_stop_words=True,
                    l_output_attrs=['Title', 'Developer', 'Publisher','Platform','Genre','Rating','ReleaseDate', 'CleanRDate'],
                   r_output_attrs=['Title', 'Developer', 'Publisher', 'Platform','Genre','Rating','ReleaseDate', 'CleanRDate'])

0% [##############################] 100% | ETA: 00:00:00
Total time elapsed: 00:00:13


In [18]:
len(C1)

979045

In [19]:
C1.head()

_id ltable_ID rtable_ID                  ltable_Title  \
0    0        a8        b1                   007 Legends   
1    1       a10        b1  007: The World is not Enough   
2    2       a11        b1      007: Tomorrow Never Dies   
3    3     a2954        b1          GoldenEye 007 (2010)   
4    4     a2955        b1       GoldenEye 007: Reloaded   

                                         ltable_Developer  \
0                          Eurocom Entertainment Software   
1  Eurocom Entertainment Software|Black Ops Entertainment   
2                                 Black Ops Entertainment   
3                  n-Space|Eurocom Entertainment Software   
4                          Eurocom Entertainment Software   

                                                  ltable_Publisher  \
0                                                       Activision   
1                                                  Electronic Arts   
2  MGM Interactive|EA Games|Electronic Arts|Electronic Arts Victor   
3                                              Activision|Nintendo   
4                                                       Activision   

                   ltable_Platform                    ltable_Genre  \
0  PC|PlayStation 3|Wii U|Xbox 360  First-Person|Shooter|3D|Action   
1          Nintendo 64|PlayStation  Action|First-Person|Shooter|3D   
2                      PlayStation  Third-Person|3D|Action|Shooter   
3                           DS|Wii  Shooter|3D|Action|First-Person   
4           PlayStation 3|Xbox 360  First-Person|Shooter|3D|Action   

   ltable_Rating ltable_ReleaseDate ltable_CleanRDate rtable_Title  \
0            5.1        16 Oct 2012        2012-10-16  007 Legends   
1            7.6         1 Nov 2000        2000-11-01  007 Legends   
2            6.3        16 Nov 1999        1999-11-16  007 Legends   
3            8.4         2 Nov 2010        2010-11-02  007 Legends   
4            7.3         1 Nov 2011        2011-11-01  007 Legends   

  rtable_Developer rtable_Publisher                  rtable_Platform  \
0          Eurocom       Activision  PC|PlayStation 3|Wii U|Xbox 360   
1          Eurocom       Activision  PC|PlayStation 3|Wii U|Xbox 360   
2          Eurocom       Activision  PC|PlayStation 3|Wii U|Xbox 360   
3          Eurocom       Activision  PC|PlayStation 3|Wii U|Xbox 360   
4          Eurocom       Activision  PC|PlayStation 3|Wii U|Xbox 360   

           rtable_Genre  rtable_Rating rtable_ReleaseDate rtable_CleanRDate  
0  First-Person|Shooter            5.0    16 October 2012        2012-10-16  
1  First-Person|Shooter            5.0    16 October 2012        2012-10-16  
2  First-Person|Shooter            5.0    16 October 2012        2012-10-16  
3  First-Person|Shooter            5.0    16 October 2012        2012-10-16  
4  First-Person|Shooter            5.0    16 October 2012        2012-10-16

### Debug Overlap blocker output

In [20]:
dbg1 = em.debug_blocker(C1, dfTableA, dfTableB)

In [21]:
dbg1

_id  similarity ltable_ID rtable_ID                    ltable_Title  \
0      0    1.000000     a1492     b1479                               D   
1      1    0.625000     a4804      b397                   Naval Warfare   
2      2    0.625000     a6844     b5805                             SSX   
3      3    0.545455     a4923       b42                NCAA GameBreaker   
4      4    0.500000     a8641     b8252                         Worms 2   
5      5    0.500000     a7710     b7501                     The Swapper   
6      6    0.500000     a8639     b1241                Worldwide Soccer   
7      7    0.500000     a1821     b1269                            DiRT   
8      8    0.461538     a7403     b6938                   The Ant Bully   
9      9    0.454545      a305     b5905                       Anomaly 2   
10    10    0.454545     a4905     b4053                NCAA Football 10   
11    11    0.454545     a3748     b7344                           Knack   
12    12    0.454545     a6213      b371                       Sanctum 2   
13    13    0.454545      a772       b42                          Blasto   
14    14    0.444444     a1758     b1497                         Deus Ex   
15    15    0.444444     a8043     b2057                      Transistor   
16    16    0.444444     a8390     b2898                        WarGames   
17    17    0.444444     a2081     b4892                       Dune 2000   
18    18    0.444444     a1802     b6646                   Dino Crisis 2   
19    19    0.444444     a6228     b4894                          Scarab   
20    20    0.444444     a1768     b5351                   Devil May Cry   
21    21    0.428571     a1161     b8329                     Chew-Man-Fu   
22    22    0.428571     a8664     b6978                   Wrecking Crew   
23    23    0.428571     a8664     b2935                   Wrecking Crew   
24    24    0.428571      a499     b1340                          Axelay   
25    25    0.416667     a5801       b42                     Rally Cross   
26    26    0.416667     a4498       b42                          MLB 99   
27    27    0.416667     a1245     b1369              Codename: Outbreak   
28    28    0.416667     a7494     b4608                 The Gunstringer   
29    29    0.416667     a7326     b5669                   Tekken Hybrid   
..   ...         ...       ...       ...                             ...   
170  170    0.333333     a8664     b3405                   Wrecking Crew   
171  171    0.333333     a8058     b2898                      Trespasser   
172  172    0.333333     a4682     b2497                    MVP Baseball   
173  173    0.333333     a4603     b1567                    Moto Racer 3   
174  174    0.333333     a6510     b5417                      SimCoaster   
175  175    0.333333     a6537     b6194                         Skate 2   
176  176    0.333333     a8043      b598                      Transistor   
177  177    0.333333     a6689     b2048                         Sorcery   
178  178    0.333333     a8642     b8252                        Worms 3D   
179  179    0.333333     a5729     b2237                            Pyre   
180  180    0.333333     a7971     b1524                           Toren   
181  181    0.333333     a8856     b5675                        Ziggurat   
182  182    0.333333     a5045     b4053                    NFL Street 3   
183  183    0.333333     a7710     b5116                     The Swapper   
184  184    0.333333     a6781      b698  Spider-Man 2: The Sinister Six   
185  185    0.333333     a8043      b334                      Transistor   
186  186    0.333333     a7606     b7153                         The Maw   
187  187    0.333333     a8043     b1212                      Transistor   
188  188    0.333333     a4937     b3274                       Necrodome   
189  189    0.333333     a8299     b6122                       Vigilance   
190  190    0.333333     a7606     b6454             

#### The number of candidates upon using overlapping blocker has reduced from 75345204 to 979045. Debugger is not showing that there are several potential matches dropped. Using overlap blocker again on candidate set. Blocking by entire keyword.

In [22]:
C2 = ob.block_candset(C1, 'Title', 'Title', word_level=True, overlap_size=1)

0% [##############################] 100% | ETA: 00:00:00
Total time elapsed: 00:00:20


In [23]:
len(C2)

530183

In [24]:
C2.head()

_id ltable_ID rtable_ID                  ltable_Title  \
0    0        a8        b1                   007 Legends   
1    1       a10        b1  007: The World is not Enough   
2    2       a11        b1      007: Tomorrow Never Dies   
3    3     a2954        b1          GoldenEye 007 (2010)   
4    4     a2955        b1       GoldenEye 007: Reloaded   

                                         ltable_Developer  \
0                          Eurocom Entertainment Software   
1  Eurocom Entertainment Software|Black Ops Entertainment   
2                                 Black Ops Entertainment   
3                  n-Space|Eurocom Entertainment Software   
4                          Eurocom Entertainment Software   

                                                  ltable_Publisher  \
0                                                       Activision   
1                                                  Electronic Arts   
2  MGM Interactive|EA Games|Electronic Arts|Electronic Arts Victor   
3                                              Activision|Nintendo   
4                                                       Activision   

                   ltable_Platform                    ltable_Genre  \
0  PC|PlayStation 3|Wii U|Xbox 360  First-Person|Shooter|3D|Action   
1          Nintendo 64|PlayStation  Action|First-Person|Shooter|3D   
2                      PlayStation  Third-Person|3D|Action|Shooter   
3                           DS|Wii  Shooter|3D|Action|First-Person   
4           PlayStation 3|Xbox 360  First-Person|Shooter|3D|Action   

   ltable_Rating ltable_ReleaseDate ltable_CleanRDate rtable_Title  \
0            5.1        16 Oct 2012        2012-10-16  007 Legends   
1            7.6         1 Nov 2000        2000-11-01  007 Legends   
2            6.3        16 Nov 1999        1999-11-16  007 Legends   
3            8.4         2 Nov 2010        2010-11-02  007 Legends   
4            7.3         1 Nov 2011        2011-11-01  007 Legends   

  rtable_Developer rtable_Publisher                  rtable_Platform  \
0          Eurocom       Activision  PC|PlayStation 3|Wii U|Xbox 360   
1          Eurocom       Activision  PC|PlayStation 3|Wii U|Xbox 360   
2          Eurocom       Activision  PC|PlayStation 3|Wii U|Xbox 360   
3          Eurocom       Activision  PC|PlayStation 3|Wii U|Xbox 360   
4          Eurocom       Activision  PC|PlayStation 3|Wii U|Xbox 360   

           rtable_Genre  rtable_Rating rtable_ReleaseDate rtable_CleanRDate  
0  First-Person|Shooter            5.0    16 October 2012        2012-10-16  
1  First-Person|Shooter            5.0    16 October 2012        2012-10-16  
2  First-Person|Shooter            5.0    16 October 2012        2012-10-16  
3  First-Person|Shooter            5.0    16 October 2012        2012-10-16  
4  First-Person|Shooter            5.0    16 October 2012        2012-10-16

In [25]:
dbg1_2 = em.debug_blocker(C2, dfTableA, dfTableB)

In [26]:
dbg1_2

_id  similarity ltable_ID rtable_ID         ltable_Title  \
0      0    1.000000     a1492     b1479                    D   
1      1    0.625000     a6844     b5805                  SSX   
2      2    0.625000     a4804      b397        Naval Warfare   
3      3    0.555556      a109      b170               ADR1FT   
4      4    0.545455     a4923       b42     NCAA GameBreaker   
5      5    0.545455       a29       b42             2 Xtreme   
6      6    0.500000     a7710     b7501          The Swapper   
7      7    0.500000     a8639     b1241     Worldwide Soccer   
8      8    0.500000     a8641     b8252              Worms 2   
9      9    0.500000     a1821     b1269                 DiRT   
10    10    0.461538     a7403     b6938        The Ant Bully   
11    11    0.454545     a4905     b4053     NCAA Football 10   
12    12    0.454545      a305     b5905            Anomaly 2   
13    13    0.454545      a772       b42               Blasto   
14    14    0.454545     a6213      b371            Sanctum 2   
15    15    0.454545     a3748     b7344                Knack   
16    16    0.444444     a8390     b2898             WarGames   
17    17    0.444444     a1768     b5351        Devil May Cry   
18    18    0.444444     a6228     b4894               Scarab   
19    19    0.444444     a2081     b4892            Dune 2000   
20    20    0.444444     a1802     b6646        Dino Crisis 2   
21    21    0.444444     a8043     b2057           Transistor   
22    22    0.444444     a1758     b1497              Deus Ex   
23    23    0.428571     a8664     b6978        Wrecking Crew   
24    24    0.428571      a499     b1340               Axelay   
25    25    0.428571     a8664     b2935        Wrecking Crew   
26    26    0.428571     a1161     b8329          Chew-Man-Fu   
27    27    0.416667     a7494     b4608      The Gunstringer   
28    28    0.416667     a4498       b42               MLB 99   
29    29    0.416667     a5801       b42          Rally Cross   
..   ...         ...       ...       ...                  ...   
170  170    0.333333     a8491     b5546      Whiplash (1996)   
171  171    0.333333     a4202     b1805  Marvel vs. Capcom 2   
172  172    0.333333     a6537     b6194              Skate 2   
173  173    0.333333     a4318     b7810           MegaRace 2   
174  174    0.333333     a8043      b334           Transistor   
175  175    0.333333     a8043     b5454           Transistor   
176  176    0.333333     a4249     b8275        MechCommander   
177  177    0.333333     a7356     b1486             Terraria   
178  178    0.333333     a5700     b7344            Puppeteer   
179  179    0.333333     a5597     b1804          Power Stone   
180  180    0.333333     a8856     b1437             Ziggurat   
181  181    0.333333     a8664     b3405        Wrecking Crew   
182  182    0.333333     a5729     b2237                 Pyre   
183  183    0.333333     a8642     b8242             Worms 3D   
184  184    0.333333     a8475     b3074   We Ski & Snowboard   
185  185    0.333333     a5615     b2464     Prince of Persia   
186  186    0.333333     a6640     b1349                 SOMA   
187  187    0.333333     a8043      b598           Transistor   
188  188    0.333333     a8827     b6684     Yu-Gi-Oh! Online   
189  189    0.333333     a6846      b553             Stacking   
190  190    0.333333     a7606     b6454              The Maw   
191  191    0.333333     a7971     b1524                Toren   
192  192    0.333333     a7606     b7153              The Maw   
193  193    0.333333     a4373     b4005               Meteos   
194  194    0.333333     a4602     b1567         Moto Racer 2   
195  195    0.333333     a8642     b8252             Worms 3D   
196  196    0.333333     a7710     b5116          The Swapper   
197  197    0.333333     a8641     b8242              Worms 2   
198  198    0.333333     a5791     b7344                 Rain   
199  199    0.333333     a8043     b1212

### Trying Attribute Equivalence Blocker

In [27]:
attrB = em.AttrEquivalenceBlocker()

In [28]:
C3 = attrB.block_tables(dfTableA, dfTableB, l_block_attr='Title', r_block_attr='Title',
                       l_output_attrs=['Title', 'Developer', 'Publisher','Platform','Genre','Rating','ReleaseDate', 'CleanRDate'],
                   r_output_attrs=['Title', 'Developer', 'Publisher', 'Platform','Genre','Rating','ReleaseDate', 'CleanRDate'])

In [29]:
len(C3)

3480

In [30]:
C3.head()

_id ltable_ID rtable_ID        ltable_Title  \
0    0        a8        b1         007 Legends   
1    1       a12       b12          1-2-Switch   
2    2       a18       b19                 140   
3    3       a25       b24  1942: Joint Strike   
4    4       a26       b23                1942   

                                                         ltable_Developer  \
0                                          Eurocom Entertainment Software   
1                                                                Nintendo   
2                                                           Jeppe Carlsen   
3                                                  Backbone Entertainment   
4  Elite Systems Ltd.|Capcom|Digital Eclipse|Micronics|Syrox Developments   

                                                 ltable_Publisher  \
0                                                      Activision   
1                                                        Nintendo   
2         Jeppe Carlsen|Double Fine Productions|Abstraction Games   
3                                                          Capcom   
4  Elite Systems Ltd.|Capcom|Elite|ASCII Entertainment|Zafiro|MCM   

                                                                                       ltable_Platform  \
0                                                                      PC|PlayStation 3|Wii U|Xbox 360   
1                                                                                      Nintendo Switch   
2                                     Linux|Macintosh|PC|PlayStation 4|PlayStation Vita|Wii U|Xbox One   
3                                                                               PlayStation 3|Xbox 360   
4  Amstrad CPC|Android|Arcade Games|Commodore 64|FM-7|Game Boy Color|iOS (iPhone/iPad)|Mobile|MSX|N...   

                     ltable_Genre  ltable_Rating ltable_ReleaseDate  \
0  First-Person|Shooter|3D|Action            5.1        16 Oct 2012   
1                  Party/Minigame            2.0         3 Mar 2017   
2            Action|Platformer|2D            7.6        16 Oct 2013   
3          Shoot-'Em-Up|2D|Action            7.2        23 Jul 2008   
4          2D|Action|Shoot-'Em-Up            7.6               1984   

  ltable_CleanRDate        rtable_Title        rtable_Developer  \
0        2012-10-16         007 Legends                 Eurocom   
1        2017-03-03          1-2-Switch            Nintendo EPD   
2        2013-10-16                 140           Jeppe Carlsen   
3        2008-07-23  1942: Joint Strike  Backbone Entertainment   
4        1984-01-01                1942                  Capcom   

  rtable_Publisher                  rtable_Platform  \
0       Activision  PC|PlayStation 3|Wii U|Xbox 360   
1         Nintendo                  Nintendo Switch   
2              NaN                     Linux|Mac|PC   
3           Capcom           PlayStation 3|Xbox 360   
4           Capcom                       Arcade|NES   

                 rtable_Genre  rtable_Rating rtable_ReleaseDate  \
0        First-Person|Shooter            5.0    16 October 2012   
1                Action|Party            5.4      03 March 2017   
2       Music/Rhythm|Platform            7.4    16 October 2013   
3  Vertical scrolling shooter            6.6       23 July 2008   
4                         NaN            6.2      December 1984   

  rtable_CleanRDate  
0        2012-10-16  
1        2017-03-03  
2        2013-10-16  
3        2008-07-23  
4        1984-12-01

In [31]:
dbg2 = em.debug_blocker(C3, dfTableA, dfTableB)

In [32]:
dbg2

_id  similarity ltable_ID rtable_ID  \
0      0    1.000000     a5169     b4890   
1      1    0.888889     a5395     b5116   
2      2    0.875000     a2831     b2810   
3      3    0.857143     a3442     b8469   
4      4    0.846154     a5692     b5369   
5      5    0.833333     a6951     b6489   
6      6    0.833333     a7265     b6884   
7      7    0.833333     a8852     b8410   
8      8    0.818182     a4416     b4432   
9      9    0.818182     a6952     b6490   
10    10    0.818182     a6953     b6492   
11    11    0.818182     a4529     b4525   
12    12    0.818182     a6398     b6066   
13    13    0.800000       a35       b41   
14    14    0.800000     a7599     b1858   
15    15    0.800000     a7934     b7657   
16    16    0.800000     a7621     b7324   
17    17    0.800000     a4417     b4436   
18    18    0.800000     a7762     b7510   
19    19    0.800000     a7994     b7691   
20    20    0.800000     a6841     b5808   
21    21    0.785714     a4206     b3034   
22    22    0.777778     a3441     b8473   
23    23    0.777778     a6033     b5707   
24    24    0.777778     a5228     b4945   
25    25    0.769231     a5389     b5109   
26    26    0.750000     a6704     b6360   
27    27    0.750000     a7513     b7173   
28    28    0.750000     a7685     b7405   
29    29    0.750000     a7512     b7171   
..   ...         ...       ...       ...   
170  170    0.600000     a5928     b5596   
171  171    0.600000     a6928     b6551   
172  172    0.600000     a1240     b1254   
173  173    0.600000     a6189     b5865   
174  174    0.600000     a8220     b7906   
175  175    0.600000     a5637     b5326   
176  176    0.600000     a7025     b6632   
177  177    0.600000     a3360     b3368   
178  178    0.600000     a6552     b6207   
179  179    0.600000     a8473     b8132   
180  180    0.590909     a1840     b2276   
181  181    0.588235     a3001     b2976   
182  182    0.588235     a4080     b4077   
183  183    0.588235     a6711     b6368   
184  184    0.588235     a3895     b3901   
185  185    0.583333       a92      b146   
186  186    0.583333      a756     b5990   
187  187    0.583333     a1524     b1511   
188  188    0.583333     a3662     b3593   
189  189    0.583333     a2536       b33   
190  190    0.583333     a3587     b3559   
191  191    0.583333     a5749     b5441   
192  192    0.583333     a5673     b5350   
193  193    0.583333     a3686     b3651   
194  194    0.583333     a2599      b107   
195  195    0.583333     a6407     b6075   
196  196    0.583333     a3087     b3063   
197  197    0.583333     a4176     b4172   
198  198    0.583333     a6901     b6521   
199  199    0.583333     a7154     b6780   

                                                 ltable_Title  \
0                                                  Not a Hero   
1                                                  Perception   
2                                             Game of Thrones   
3                                                    inFamous   
4                          Psychonauts in The Rhombus of Ruin   
5                            Starcraft II: Heart of the Swarm   
6                       Tales From Space: Mutant Blobs Attack   
7                                                     ZENONIA   
8                              Middle-earth: Shadow of Mordor   
9                            Starcraft II: Legacy of the Void   
10                             Starcraft II: Wings of Liberty   
11                               Monaco: What's Yours Is Mine   
12                        Sherlock Holmes vs. Jack the Ripper   
13                                                 25 to Life   
14                    The Magical Quest Starring Mickey Mouse   
15                                  Tony Hawk's Pro Skater 2x   
16                                         The Next Big Thing   
17                                Middle-earth: Shadow of War   
18                                       

#### Attribute equivalence blocker generates candidate set with almost all positives but since it is case sensitive we are missing on lot of positive samples.

#### Trying Rule based blocking on Candidate set C2 generated after running Overlap blocker by q-gram and word overlap

In [38]:
block_f = em.get_features_for_blocking(dfTableA, dfTableB, validate_inferred_attr_types=False)

In [41]:
block_f

feature_name left_attribute  \
0                                ID_ID_lev_dist             ID   
1                                 ID_ID_lev_sim             ID   
2                                     ID_ID_jar             ID   
3                                     ID_ID_jwn             ID   
4                                     ID_ID_exm             ID   
5                         ID_ID_jac_qgm_3_qgm_3             ID   
6                   Title_Title_jac_qgm_3_qgm_3          Title   
7               Title_Title_cos_dlm_dc0_dlm_dc0          Title   
8               Title_Title_jac_dlm_dc0_dlm_dc0          Title   
9                               Title_Title_mel          Title   
10                         Title_Title_lev_dist          Title   
11                          Title_Title_lev_sim          Title   
12                              Title_Title_nmw          Title   
13                               Title_Title_sw          Title   
14          Developer_Developer_jac_qgm_3_qgm_3      Developer   
15      Developer_Developer_cos_dlm_dc0_dlm_dc0      Developer   
16      Developer_Developer_jac_dlm_dc0_dlm_dc0      Developer   
17                      Developer_Developer_mel      Developer   
18                 Developer_Developer_lev_dist      Developer   
19                  Developer_Developer_lev_sim      Developer   
20                      Developer_Developer_nmw      Developer   
21                       Developer_Developer_sw      Developer   
22          Publisher_Publisher_jac_qgm_3_qgm_3      Publisher   
23      Publisher_Publisher_cos_dlm_dc0_dlm_dc0      Publisher   
24      Publisher_Publisher_jac_dlm_dc0_dlm_dc0      Publisher   
25                      Publisher_Publisher_mel      Publisher   
26                 Publisher_Publisher_lev_dist      Publisher   
27                  Publisher_Publisher_lev_sim      Publisher   
28                      Publisher_Publisher_nmw      Publisher   
29                       Publisher_Publisher_sw      Publisher   
..                                          ...            ...   
34                   Platform_Platform_lev_dist       Platform   
35                    Platform_Platform_lev_sim       Platform   
36                        Platform_Platform_nmw       Platform   
37                         Platform_Platform_sw       Platform   
38                  Genre_Genre_jac_qgm_3_qgm_3          Genre   
39              Genre_Genre_cos_dlm_dc0_dlm_dc0          Genre   
40              Genre_Genre_jac_dlm_dc0_dlm_dc0          Genre   
41                              Genre_Genre_mel          Genre   
42                         Genre_Genre_lev_dist          Genre   
43                          Genre_Genre_lev_sim          Genre   
44                              Genre_Genre_nmw          Genre   
45                               Genre_Genre_sw          Genre   
46      ReleaseDate_ReleaseDate_jac_qgm_3_qgm_3    ReleaseDate   
47  ReleaseDate_ReleaseDate_cos_dlm_dc0_dlm_dc0    ReleaseDate   
48  ReleaseDate_ReleaseDate_jac_dlm_dc0_dlm_dc0    ReleaseDate   
49                  ReleaseDate_ReleaseDate_mel    ReleaseDate   
50             ReleaseDate_ReleaseDate_lev_dist    ReleaseDate   
51              ReleaseDate_ReleaseDate_lev_sim    ReleaseDate   
52                  ReleaseDate_ReleaseDate_nmw    ReleaseDate   
53                   ReleaseDate_ReleaseDate_sw    ReleaseDate   
54                            Rating_Rating_exm         Rating   
55                            Rating_Rating_anm         Rating   
56                       Rating_Rating_lev_dist         Rating   
57                        Rating_Rating_lev_sim         Rating   
58               CleanRDate_CleanRDate_lev_dist     CleanRDate   
59                CleanRDate_CleanRDate_lev_sim     CleanRDate   
60                    CleanRDate_CleanRDate_jar     CleanRDate   
61                    CleanRDate_CleanRDate_jwn     CleanRDate   
62                    CleanRDate_CleanRDate_exm     CleanRDate   
63       

#### Add similarity measure as rule for candidate in set C2

In [175]:
rb = em.RuleBasedBlocker()
rb.add_rule(['Title_Title_lev_sim(ltuple, rtuple) < 0.4'], block_f)

'_rule_0'

In [176]:
C4 = rb.block_candset(C2)

0% [##############################] 100% | ETA: 00:00:00
Total time elapsed: 00:01:14


In [177]:
len(C4)

70668

In [178]:
C4.head()

_id ltable_ID rtable_ID        ltable_Title  \
0     0        a8        b1         007 Legends   
8     8      a973        b1     Burnout Legends   
14   14     a2557        b1    Fighting Legends   
15   15     a3826        b1   League of Legends   
24   24     a4295        b1  Mega Man Legends 2   

                  ltable_Developer                    ltable_Publisher  \
0   Eurocom Entertainment Software                          Activision   
8    Visual Impact|Criterion Games            EA Games|Electronic Arts   
14                Maximum Charisma                    Maximum Charisma   
15                      Riot Games  TransGaming Inc.|Riot Games|Garena   
24                          Capcom                              Capcom   

                    ltable_Platform                    ltable_Genre  \
0   PC|PlayStation 3|Wii U|Xbox 360  First-Person|Shooter|3D|Action   
8                            DS|PSP           Arcade|Driving/Racing   
14                               PC                    Role-Playing   
15                     Macintosh|PC      Free-to-Play|MOBA|Strategy   
24               PC|PlayStation|PSP             Action|Role-Playing   

    ltable_Rating ltable_ReleaseDate ltable_CleanRDate rtable_Title  \
0             5.1        16 Oct 2012        2012-10-16  007 Legends   
8             8.7        13 Sep 2005        2005-09-13  007 Legends   
14            6.6        31 Oct 2001        2001-10-31  007 Legends   
15            8.0        27 Oct 2009        2009-10-27  007 Legends   
24            8.7        24 Oct 2000        2000-10-24  007 Legends   

   rtable_Developer rtable_Publisher                  rtable_Platform  \
0           Eurocom       Activision  PC|PlayStation 3|Wii U|Xbox 360   
8           Eurocom       Activision  PC|PlayStation 3|Wii U|Xbox 360   
14          Eurocom       Activision  PC|PlayStation 3|Wii U|Xbox 360   
15          Eurocom       Activision  PC|PlayStation 3|Wii U|Xbox 360   
24          Eurocom       Activision  PC|PlayStation 3|Wii U|Xbox 360   

            rtable_Genre  rtable_Rating rtable_ReleaseDate rtable_CleanRDate  
0   First-Person|Shooter            5.0    16 October 2012        2012-10-16  
8   First-Person|Shooter            5.0    16 October 2012        2012-10-16  
14  First-Person|Shooter            5.0    16 October 2012        2012-10-16  
15  First-Person|Shooter            5.0    16 October 2012        2012-10-16  
24  First-Person|Shooter            5.0    16 October 2012        2012-10-16

In [179]:
dbg1_3 = em.debug_blocker(C4, dfTableA, dfTableB)

In [180]:
dbg1_3

_id  similarity ltable_ID rtable_ID                  ltable_Title  \
0      0    1.000000     a1492     b1479                             D   
1      1    0.857143     a3442     b8469                      inFamous   
2      2    0.833333     a8852     b8410                       ZENONIA   
3      3    0.714286     a2201     b2185                          Elex   
4      4    0.714286     a6134     b5462                        Ruiner   
5      5    0.714286     a6218     b5909                      SATAZIUS   
6      6    0.666667     a3902     b3913                         LIMBO   
7      7    0.666667     a2491     b2497                          FATE   
8      8    0.666667     a3460     b3463                        INSIDE   
9      9    0.625000     a2334     b2329                    EVERYTHING   
10    10    0.625000     a4576     b4560                   Moon (2009)   
11    11    0.625000     a6844     b5805                           SSX   
12    12    0.625000     a5610     b5297                          Prey   
13    13    0.625000     a4804      b397                 Naval Warfare   
14    14    0.600000     a8473     b8132                 WE ARE DOOMED   
15    15    0.583333     a2536       b33  FIFA World Cup: Germany 2006   
16    16    0.571429      a385      b437                          ARMS   
17    17    0.571429     a1212     b1230                          Claw   
18    18    0.571429     a5113     b4826                          NIER   
19    19    0.555556     a7535     b3654       The King of Fighters XI   
20    20    0.555556     a1927     b1924                          DOOM   
21    21    0.555556      a109      b170                        ADR1FT   
22    22    0.545455       a29       b42                      2 Xtreme   
23    23    0.545455     a4923       b42              NCAA GameBreaker   
24    24    0.500000     a8639     b1241              Worldwide Soccer   
25    25    0.500000     a6151     b5773                          Rune   
26    26    0.500000     a6640     b6285                          SOMA   
27    27    0.500000     a1821     b1269                          DiRT   
28    28    0.500000     a8641     b8252                       Worms 2   
29    29    0.500000     a7710     b7501                   The Swapper   
..   ...         ...       ...       ...                           ...   
170  170    0.363636     a8439     b2164            Warriors All-Stars   
171  171    0.363636     a5791     b5487                          Rain   
172  172    0.363636     a4604     b1567                    Moto Racer   
173  173    0.363636     a5046     b4056                    NFL Street   
174  174    0.363636     a4297     b4313    Mega Man Maverick Hunter X   
175  175    0.363636     a3938     b6188          LocoRoco Cocoreccho!   
176  176    0.363636     a3937     b5373                    LocoRoco 2   
177  177    0.363636     a6846     b3479                      Stacking   
178  178    0.363636     a8843     b1324                       Zathura   
179  179    0.363636     a7195     b5731            Supreme Ruler 2010   
180  180    0.363636     a7842     b5908       TNN Outdoors Pro Hunter   
181  181    0.363636     a6804     b3945                 'Splosion Man   
182  182    0.363636     a6646     b6314                 Sonic Advance   
183  183    0.363636     a7531     b3654      The King of Fighters '99   
184  184    0.363636     a4868     b2391                 NBA Live 2005   
185  185    0.363636     a7601     b7031                      The Mark   
186  186    0.363636     a5366     b3947                     Patapon 2   
187  187    0.363636     a5047     b4052                      NFL Tour   
188  188    0.363636     a4858     b4678                   NBA Live 10   
189  189    0.363636     a5058     b2396                        NHL 13   
190  190    0.363636     a5395     b7423                    Perception   
191  191    0.363636     a8439     b2169            Warriors All-Stars   
192  192 

#### A few single word titles are blocked mainly because the blocking is case-sensitive to title attribute. A case-insensitive version will help reduce the number of dropped matches.

Reducing the candidate blockset further by using black box matcher with custom function to check for release date similarity.

In [110]:
# release date compare function
# if titles release match within a year in both tables we don't block it
def compareReleaseDate(A, B):
    A_cleanRDate = A['CleanRDate']
    B_cleanRDate = B['CleanRDate']
    
    try:
        d1 = datetime.strptime(A_cleanRDate, "%Y-%m-%d")
        d2 = datetime.strptime(B_cleanRDate, "%Y-%m-%d")
        dayDiff = abs(d2-d1).days

        if dayDiff > 366:
            return True
    except:
        return False

In [111]:
bb = em.BlackBoxBlocker()
# Set the black box function
bb.set_black_box_function(compareReleaseDate)
C5 = bb.block_candset(C4)

0% [##############################] 100% | ETA: 00:00:00
Total time elapsed: 00:00:07


In [112]:
len(C5)

13062

### Combine candidates

In [181]:
# combining true matches as found by Attribute equivalence blocker
C = em.combine_blocker_outputs_via_union([C3, C5])
len(C)

13186

In [205]:
# save candidate set C to file
C.to_csv('selectedCandidates.csv',index=False)

In [167]:
dbg1_4 = em.debug_blocker(C, dfTableA, dfTableB)

In [169]:
dbg1_4

_id  similarity ltable_ID rtable_ID  \
0      0    0.857143     a3442     b8469   
1      1    0.833333     a8852     b8410   
2      2    0.750000     a7512     b7171   
3      3    0.714286     a6218     b5909   
4      4    0.714286     a6134     b5462   
5      5    0.714286     a2201     b2185   
6      6    0.666667     a2781     b2763   
7      7    0.666667     a2491     b2497   
8      8    0.666667     a3902     b3913   
9      9    0.666667     a3460     b3463   
10    10    0.642857     a1989     b2015   
11    11    0.625000     a4576     b4560   
12    12    0.625000     a5610     b5297   
13    13    0.625000     a4804      b397   
14    14    0.625000     a6844     b5805   
15    15    0.625000     a2334     b2329   
16    16    0.600000     a7025     b6632   
17    17    0.600000     a8473     b8132   
18    18    0.588235     a4080     b4077   
19    19    0.583333     a6407     b6075   
20    20    0.583333     a2536       b33   
21    21    0.571429     a7513     b7174   
22    22    0.571429     a5113     b4826   
23    23    0.571429     a1212     b1230   
24    24    0.571429      a385      b437   
25    25    0.562500     a1990     b2016   
26    26    0.555556     a4080     b4079   
27    27    0.555556     a4080     b4078   
28    28    0.555556      a109      b170   
29    29    0.555556     a4080     b4081   
..   ...         ...       ...       ...   
170  170    0.444444     a1802     b6646   
171  171    0.444444     a3012     b2988   
172  172    0.444444     a3013     b2992   
173  173    0.444444     a1758     b1497   
174  174    0.444444     a1768     b5351   
175  175    0.444444     a6228     b4894   
176  176    0.444444     a4307     b4285   
177  177    0.444444     a2768     b2762   
178  178    0.444444     a7894     b7622   
179  179    0.444444     a8043     b2057   
180  180    0.444444     a5597     b5287   
181  181    0.444444     a7023     b6630   
182  182    0.444444     a2081     b4892   
183  183    0.444444     a2011     b2033   
184  184    0.444444     a8390     b2898   
185  185    0.444444     a7810     b7554   
186  186    0.444444     a6411     b6068   
187  187    0.444444     a4560     b4546   
188  188    0.437500     a1986     b2013   
189  189    0.437500     a1506     b1502   
190  190    0.437500     a4144     b4171   
191  191    0.437500     a6145     b5775   
192  192    0.437500     a1985     b2001   
193  193    0.437500     a8394     b8078   
194  194    0.437500     a8395     b8081   
195  195    0.437500     a6406     b6074   
196  196    0.437500     a7566     b7253   
197  197    0.437500     a8743     b6574   
198  198    0.437500     a7871     b7602   
199  199    0.428571     a8664     b6978   

                                                 ltable_Title  \
0                                                    inFamous   
1                                                     ZENONIA   
2                The Incredible Adventures of Van Helsing III   
3                                                    SATAZIUS   
4                                                      Ruiner   
5                                                        Elex   
6                    Fullmetal Alchemist and the Broken Angel   
7                                                        FATE   
8                                                       LIMBO   
9                                                      INSIDE   
10                       Dragon Ball Z: The Legacy of Goku II   
11                                                Moon (2009)   
12                                                       Prey   
13                                              Naval Warfare   
14                                                        SSX   
15                                                 EVERYTHING   
16      Street Fighter III: 3rd Strike - Fight for the Future   
17                                              WE ARE DOOMED   
18   Magic: The Gathering - Duels of the 

#### Candidate set is now sufficiently reduced to get sufficient positive samples while labling for matcher to learn.

### Final blocking sequence

In [ ]:
##### A, B ----- overlap blocker (3-gram, overlap=4) 
#                                  -----> overlap blocker (1-word) 
#                                             -----> Rule Based Blocker (Title_Title_Sim < 0.4) 
#                                                                        -----> Black Box Blocker - C5
#                                                                                                       | --> C
#     A, B ------ Attr Eq Blocker (Title) --------------------------------------------------------- C3

### Sampling and Labeling

#### Select 400 tuples for labeling to be used by matcher algorithm

In [183]:
S1 = em.sample_table(C, 400)
G1 = em.label_table(S1,'gold_label')
G1.to_csv('toLabel.csv', index=False)

Column name (gold_label) is not present in dataframe


#### Tuples were labeled offline and read back here for further processing

In [113]:
G = em.read_csv_metadata('../data/Labelled.csv', 
                         key='_id',
                         ltable=dfTableA, rtable=dfTableB, 
                         fk_ltable='ltable_ID', fk_rtable='rtable_ID')
len(G)

Metadata file is not present in the given path; proceeding to read the csv file.


400

### Splitting labeled dataset into development and evaluation set

In [186]:
# Split S into development set (I) and evaluation set (J)
IJ = em.split_train_test(G, train_proportion=0.7, random_state=0)
I = IJ['train']
J = IJ['test']
print ('Development Set(I) size = ' + str(len(I)))
print ('Evaluation Set(J) size = ' + str(len(J)))

Development Set(I) size = 280
Evaluation Set(J) size = 120


In [206]:
# save set I,J to file
I.to_csv('setI.csv', index=False)
J.to_csv('setJ.csv', index=False)

### Instantiate matcher objects

In [115]:
# Create a set of ML-matchers
dt = em.DTMatcher(name='DecisionTree', random_state=0)
svm = em.SVMMatcher(name='SVM', random_state=0)
rf = em.RFMatcher(name='RF', random_state=0)
lg = em.LogRegMatcher(name='LogReg', random_state=0)
ln = em.LinRegMatcher(name='LinReg')
nb = em.NBMatcher(name='NaiveBayes')

### Generate feature set for given data

In [116]:
# Generate features
feature_table = em.get_features_for_matching(dfTableA, dfTableB, validate_inferred_attr_types=False)

In [117]:
# List the names of the features generated
feature_table['feature_name']

0                                  ID_ID_lev_dist
1                                   ID_ID_lev_sim
2                                       ID_ID_jar
3                                       ID_ID_jwn
4                                       ID_ID_exm
5                           ID_ID_jac_qgm_3_qgm_3
6                     Title_Title_jac_qgm_3_qgm_3
7                 Title_Title_cos_dlm_dc0_dlm_dc0
8                 Title_Title_jac_dlm_dc0_dlm_dc0
9                                 Title_Title_mel
10                           Title_Title_lev_dist
11                            Title_Title_lev_sim
12                                Title_Title_nmw
13                                 Title_Title_sw
14            Developer_Developer_jac_qgm_3_qgm_3
15        Developer_Developer_cos_dlm_dc0_dlm_dc0
16        Developer_Developer_jac_dlm_dc0_dlm_dc0
17                        Developer_Developer_mel
18                   Developer_Developer_lev_dist
19                    Developer_Developer_lev_sim


### Convert development set I into feature vectors

In [118]:
# Convert the I into a set of feature vectors using F
H = em.extract_feature_vecs(I, 
                            feature_table=feature_table, 
                            attrs_after='gold_label',
                            show_progress=False)

In [119]:
H.head()

_id ltable_ID rtable_ID  ID_ID_lev_dist  ID_ID_lev_sim  ID_ID_jar  \
92   913082     a7987     b7683               3            0.4   0.600000   
223  231382     a1983     b2008               5            0.0   0.466667   
234  304609     a2588     b2606               4            0.2   0.466667   
232  298331     a1418     b2571               5            0.0   0.466667   
377  771435      a843     b6710               5            0.0   0.000000   

     ID_ID_jwn  ID_ID_exm  ID_ID_jac_qgm_3_qgm_3  Title_Title_jac_qgm_3_qgm_3  \
92    0.600000          0                    0.0                     0.157895   
223   0.466667          0                    0.0                     1.000000   
234   0.466667          0                    0.0                     0.233333   
232   0.466667          0                    0.0                     0.361111   
377   0.000000          0                    0.0                     0.307692   

        ...      Rating_Rating_anm  Rating_Rating_lev_dist  \
92      ...               0.365854                     2.0   
223     ...               0.894118                     2.0   
234     ...               0.692308                     2.0   
232     ...               0.879121                     2.0   
377     ...               0.908046                     2.0   

     Rating_Rating_lev_sim  CleanRDate_CleanRDate_lev_dist  \
92                0.333333                             4.0   
223               0.333333                             0.0   
234               0.333333                             NaN   
232               0.333333                             NaN   
377               0.333333                             NaN   

     CleanRDate_CleanRDate_lev_sim  CleanRDate_CleanRDate_jar  \
92                             0.6                        0.8   
223                            1.0                        1.0   
234                            NaN                        NaN   
232                            NaN                        NaN   
377                            NaN                        NaN   

     CleanRDate_CleanRDate_jwn  CleanRDate_CleanRDate_exm  \
92                        0.86                        0.0   
223                       1.00                        1.0   
234                        NaN                        NaN   
232                        NaN                        NaN   
377                        NaN                        NaN   

     CleanRDate_CleanRDate_jac_qgm_3_qgm_3  gold_label  
92                                0.142857           0  
223                               1.000000           1  
234                                    NaN           0  
232                                    NaN           0  
377                                    NaN           0  

[5 rows x 68 columns]

### Fill missing feature value computation (NaN) with relevant values

In [120]:
#set all missing value NaN to appropriate numeric values
import math
for idx, row in H.iterrows():
    for column in H.columns:
        v = row[column]
        if isinstance(v, int) or isinstance(v, float):
            if math.isnan(v):
                if column.find('lev_dist') != -1:
                    H.set_value(idx, column, 5)
                else:
                    H.set_value(idx, column, 0)

### Determine best learning algorithm using cross-validation

In [121]:
# Select the best ML matcher using 5-fold CV
result = em.select_matcher([dt, rf, svm, ln, lg, nb], table=H, 
        exclude_attrs=['_id', 'ltable_ID', 'rtable_ID', 'gold_label'],
        k=5,
        target_attr='gold_label', metrics_to_display=['precision', 'recall', 'f1'], random_state=0)
result['cv_stats']

Matcher  Average precision  Average recall  Average f1
0  DecisionTree           0.965000        0.969318    0.966396
1            RF           1.000000        0.962500    0.979310
2           SVM           0.000000        0.000000    0.000000
3        LinReg           1.000000        0.936667    0.964532
4        LogReg           0.931587        0.944318    0.936896
5    NaiveBayes           0.924373        0.969318    0.944382

### Precision Stats for each fold of CV

In [122]:
result['drill_down_cv_stats']['precision']

Name  \
0  DecisionTree   
1            RF   
2           SVM   
3        LinReg   
4        LogReg   
5    NaiveBayes   

                                                                                Matcher  \
0          <py_entitymatching.matcher.dtmatcher.DTMatcher object at 0x0000020217776710>   
1          <py_entitymatching.matcher.rfmatcher.RFMatcher object at 0x00000202177764E0>   
2        <py_entitymatching.matcher.svmmatcher.SVMMatcher object at 0x0000020217776390>   
3  <py_entitymatching.matcher.linregmatcher.LinRegMatcher object at 0x0000020217776470>   
4  <py_entitymatching.matcher.logregmatcher.LogRegMatcher object at 0x0000020217776F98>   
5          <py_entitymatching.matcher.nbmatcher.NBMatcher object at 0x0000020217776128>   

   Num folds  Fold 1    Fold 2    Fold 3    Fold 4    Fold 5  Mean score  
0          5  0.9375  0.950000  1.000000  1.000000  0.937500    0.965000  
1          5  1.0000  1.000000  1.000000  1.000000  1.000000    1.000000  
2          5  0.0000  0.000000  0.000000  0.000000  0.000000    0.000000  
3          5  1.0000  1.000000  1.000000  1.000000  1.000000    1.000000  
4          5  1.0000  0.950000  0.888889  0.952381  0.866667    0.931587  
5          5  0.9375  0.904762  0.842105  1.000000  0.937500    0.924373

### Random Forest get 100% precision at first trial of CV
### Using Random Forest classifier for testing evaluate set J

In [123]:
# Train using feature vectors from I 
rf.fit(table=H, 
       exclude_attrs=['_id', 'ltable_ID', 'rtable_ID', 'gold_label'], 
       target_attr='gold_label')

### Convert evaluation set J into feature vectors and fill missing values

In [124]:
# Convert J into a set of feature vectors using feature table
L = em.extract_feature_vecs(J, feature_table=feature_table,
                            attrs_after='gold_label', show_progress=False)

#set all missing value NaN to appropriate numeric values
for idx, row in L.iterrows():
    for column in L.columns:
        v = row[column]
        if isinstance(v, int) or isinstance(v, float):
            if math.isnan(v):
                if column.find('lev_dist') != -1:
                    L.set_value(idx, column, 5)
                else:
                    L.set_value(idx, column, 0)

### Predicting the matches

In [125]:
# Predict on L 
predictions = rf.predict(table=L, exclude_attrs=['_id', 'ltable_ID', 'rtable_ID', 'gold_label'], 
              append=True, target_attr='predicted', inplace=False)

### Evaluating the predictions

In [190]:
# Evaluate the predictions
print ('Random Forst prediction summary:')
eval_result = em.eval_matches(predictions, 'gold_label', 'predicted')
em.print_eval_summary(eval_result)

Random Forst prediction summary:
Precision : 96.43% (27/28)
Recall : 84.38% (27/32)
F1 : 90.0%
False positives : 1 (out of 28 positive predictions)
False negatives : 5 (out of 92 negative predictions)


In [211]:
JStore = J.copy()
JStore['predictedVal'] = pd.Series()
for idx, row in predictions.iterrows():
    JStore.set_value(idx, 'predictedVal', row['predicted'])
JStore.to_csv('RFprediction.csv', index=False)

# Following section is for completeness of project report

## Evaluating other matchers for Evaluation Set

In [198]:
dt.fit(table=H, 
       exclude_attrs=['_id', 'ltable_ID', 'rtable_ID', 'gold_label'], 
       target_attr='gold_label')

svm.fit(table=H, 
       exclude_attrs=['_id', 'ltable_ID', 'rtable_ID', 'gold_label'], 
       target_attr='gold_label')

lg.fit(table=H, 
       exclude_attrs=['_id', 'ltable_ID', 'rtable_ID', 'gold_label'], 
       target_attr='gold_label')

ln.fit(table=H, 
       exclude_attrs=['_id', 'ltable_ID', 'rtable_ID', 'gold_label'], 
       target_attr='gold_label')

nb.fit(table=H, 
       exclude_attrs=['_id', 'ltable_ID', 'rtable_ID', 'gold_label'], 
       target_attr='gold_label')

In [199]:
predictionsDT = dt.predict(table=L, exclude_attrs=['_id', 'ltable_ID', 'rtable_ID', 'gold_label'], 
              append=True, target_attr='predicted', inplace=False)

predictionsSVM = svm.predict(table=L, exclude_attrs=['_id', 'ltable_ID', 'rtable_ID', 'gold_label'], 
              append=True, target_attr='predicted', inplace=False)

predictionsLogR = lg.predict(table=L, exclude_attrs=['_id', 'ltable_ID', 'rtable_ID', 'gold_label'], 
              append=True, target_attr='predicted', inplace=False)

predictionsLinR = ln.predict(table=L, exclude_attrs=['_id', 'ltable_ID', 'rtable_ID', 'gold_label'], 
              append=True, target_attr='predicted', inplace=False)

predictionsNB = nb.predict(table=L, exclude_attrs=['_id', 'ltable_ID', 'rtable_ID', 'gold_label'], 
              append=True, target_attr='predicted', inplace=False)

In [200]:
print ('Decision Tree prediction summary:')
eval_result = em.eval_matches(predictionsDT, 'gold_label', 'predicted')
em.print_eval_summary(eval_result)

Decision Tree prediction summary:
Precision : 90.32% (28/31)
Recall : 87.5% (28/32)
F1 : 88.89%
False positives : 3 (out of 31 positive predictions)
False negatives : 4 (out of 89 negative predictions)


In [201]:
print ('SVM prediction summary:')
eval_result = em.eval_matches(predictionsSVM, 'gold_label', 'predicted')
em.print_eval_summary(eval_result)

SVM prediction summary:
Precision : 0.0% (0/0)
Recall : 0.0% (0/32)
F1 : 0.0%
False positives : 0 (out of 0 positive predictions)
False negatives : 32 (out of 120 negative predictions)


In [202]:
print ('Logistic Regression prediction summary:')
eval_result = em.eval_matches(predictionsLogR, 'gold_label', 'predicted')
em.print_eval_summary(eval_result)

Logistic Regression prediction summary:
Precision : 90.91% (30/33)
Recall : 93.75% (30/32)
F1 : 92.31%
False positives : 3 (out of 33 positive predictions)
False negatives : 2 (out of 87 negative predictions)


In [203]:
print ('Linear Regression prediction summary:')
eval_result = em.eval_matches(predictionsLinR, 'gold_label', 'predicted')
em.print_eval_summary(eval_result)

Linear Regression prediction summary:
Precision : 93.1% (27/29)
Recall : 84.38% (27/32)
F1 : 88.52%
False positives : 2 (out of 29 positive predictions)
False negatives : 5 (out of 91 negative predictions)


In [204]:
print ('Naive Bayes prediction summary:')
eval_result = em.eval_matches(predictionsNB, 'gold_label', 'predicted')
em.print_eval_summary(eval_result)

Naive Bayes prediction summary:
Precision : 88.57% (31/35)
Recall : 96.88% (31/32)
F1 : 92.54%
False positives : 4 (out of 35 positive predictions)
False negatives : 1 (out of 85 negative predictions)
